processor/src/operations/crypto_ops.rs

```diff
@@ -8,35 +8,45 @@ use super::{ExecutionError, Process};
 impl Process {
     // HASHING OPERATIONS
     // --------------------------------------------------------------------------------------------
-    /// Applies Rescue Prime permutation to the top 12 elements of the stack. The outer part of the
-    /// state is assumed to be at the top of the stack.
+    /// Applies Rescue Prime permutation to the top 12 elements of the stack. The stack is assumed
+    /// to be arranged so that the 8 elements of the rate are at the top of the stack. The capacity
+    /// word follows, with the number of elements to be hashed at the deepest position in stack[11].
+    /// For a Rescue Prime permutation of [A, B, C] where A is the capacity, the stack should be
+    /// arranged (from the top) as [C, B, A, ...].
     ///
     /// # Errors
     /// Returns an error if the stack contains fewer than 12 elements.
     pub(super) fn op_rpperm(&mut self) -> Result<(), ExecutionError> {
         self.stack.check_depth(12, "RPPERM")?;
 
         let input_state = [
-            self.stack.get(0),
-            self.stack.get(1),
-            self.stack.get(2),
-            self.stack.get(3),
-            self.stack.get(4),
-            self.stack.get(5),
-            self.stack.get(6),
-            self.stack.get(7),
-            self.stack.get(8),
-            self.stack.get(9),
-            self.stack.get(10),
             self.stack.get(11),
+            self.stack.get(10),
+            self.stack.get(9),
+            self.stack.get(8),
+            self.stack.get(7),
+            self.stack.get(6),
+            self.stack.get(5),
+            self.stack.get(4),
+            self.stack.get(3),
+            self.stack.get(2),
+            self.stack.get(1),
+            self.stack.get(0),
         ];
 
         let (_addr, output_state) = self.hasher.permute(input_state);
 
-        for (i, &value) in output_state.iter().enumerate() {
-            self.stack.set(i, value);
-        }
-        self.stack.copy_state(12);
+        self.stack.set(0, output_state[11]);
+        self.stack.set(1, output_state[10]);
+        self.stack.set(2, output_state[9]);
+        self.stack.set(3, output_state[8]);
+        self.stack.set(4, output_state[7]);
+        self.stack.set(5, output_state[6]);
+        self.stack.set(6, output_state[5]);
+        self.stack.set(7, output_state[4]);
+        // keep the capacity portion of the stack unchanged
```

Comment:
Here we should copy the entire state back to the stack. The reason is that capacity elements do in fact change as a result of the permutation.

Capacity element remain the same between two different permutations when absorbing more elements that can be processed with a single permutation, but that's not relevant here, since this operation deals with a single permutation only.

Reply:
> Capacity element remain the same between two different permutations when absorbing more elements that can be processed with a single permutation

Hm, I definitely have some confusion here in terms of how this combines with your previous note, because I thought that's what you were saying with this:

> The overall reason for having the capacity portion deep in the stack is that when we want to hash more than 8 elements, we don't need to touch the capacity portion. ... And then we'd be able to executed ```RPPERM``` immediately after this again to apply another permutation.

Combined with this, I had thought that we weren't using the capacity portion of the output anywhere, so there would be no reason we wouldn't want to preserve the capacity to allow for hashing more than 8 elements as in my understanding of  your example.

Can you please clarify for me? Did I misunderstand the previous comment? Or is this a case where we'd want to create/use a different processor operation to do hashes of more than 8 elements? What do we do with the capacity portion of the output that we're now preserving?

If I were to execute your previous example by sending inputs directly to the Winterfell hash_elements function, what would that look like? (That might clarify things for me)

Reply:
I think an example can help here. Let's say we want to hash a sequence of 16 field elements: [A, B, C, D]. Since we can absorb only 8 elements per permutation, we'll need to perform two permutations (i.e., execute `RPPERM` operation twice). Here how it will look like:

First, we need to load the stack with the first 8 elements, setting capacity portion to 0's except for the element containing total number of elements to be hashed. The stack would look like so:
```
[B, A, 0, 0, 0, 16, ...]
```
Then we execute `RPPERM`. The output of `RPPERM` is 12 field elements (3 words), let's call them T0, T1, T2. Here T0 are the capacity elements, but the values of these elements are not the same as at the start. So, the stack after `RPPERM` would look like:
```
[T2, T1, T0, ...]
```
Then, we need to "absorb" words C and D into the state. This is done by adding them (element-wise) to the rate portion. I'm going to skip all stack manipulation instructions here, but the end result should look like:
```
[T2 + D, T1 + C, T0, ...]
```
So, the above process affected only the rate portion, and the capacity portion remained untouched. Then, we execute `RPPERM` again, and get 12 more elements as output. Let's call them: R0, R1, R2. On the stack they will look like so:
```
[R2, R1, R0, ...]
```
The result of hour hash is R1. But if we wanted to continue hashing, we'd again absorb the next 8 elements into the state by adding them to R1 and R2, and again, leaving the capacity portion (which is R0) untouched.

Reply:
> of course, if we have more than 16 elements, we should initialize the first state accordingly in the very beginning

I was just going to respond to double-check this, but you got in with the ninja edit. The example makes it clear now. Thanks!